x/leverage/keeper/keeper.go

```diff
@@ -270,3 +270,166 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return err
+	}
+
+	// Repayment cannot exceed liquidator's available balance
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	if repayment.Amount.GTE(liquidatorBalance.Amount) {
+		repayment.Amount = liquidatorBalance.Amount
+	}
+
+	// Repayment cannot exceed borrower's borrowed amount of selected denom
+	if repayment.Amount.GTE(borrowed.AmountOf(repayment.Denom)) {
+		repayment.Amount = borrowed.AmountOf(repayment.Denom)
+	}
+
+	// Repayment cannot exceed borrowed value * close factor
+	repayValue, err := k.GetPrice(ctx, repayment)
+	if err != nil {
+		return err
+	}
+	if repayValue.GTE(borrowValue.Mul(closeFactor)) {
+		maxRepayValue := borrowValue.Mul(closeFactor)
+		// repayment *= (maxRepayValue / repayValue)
+		repayment.Amount = repayment.Amount.ToDec().Mul(maxRepayValue).Quo(repayValue).TruncateInt()
+		repayValue = maxRepayValue
+	}
+
+	// Given repay denom and amount, use oracle to find equivalent amount of rewardDenom's base asset
+	baseReward, err := k.EquivalentValue(ctx, repayment, baseRewardDenom)
+	if err != nil {
+		return err
+	}
+
+	// Convert reward tokens back to uTokens
+	reward, err := k.ExchangeTokens(ctx, baseReward)
+	if err != nil {
+		return err
+	}
+
+	// Apply liquidation incentive
+	reward.Amount = reward.Amount.ToDec().Mul(sdk.OneDec().Add(liquidationIncentive)).TruncateInt()
+
+	// Reward amount cannot exceed available collateral
+	if reward.Amount.GTE(collateral.AmountOf(rewardDenom)) {
+		// only pay what can be correctly compensated
+		repayment.Amount = repayment.Amount.Mul(collateral.AmountOf(rewardDenom)).Quo(reward.Amount)
+		// use all collateral of selected denom
+		reward.Amount = collateral.AmountOf(rewardDenom)
+	}
+
+	// Send repayment to leverage module account
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(
+		ctx, liquidatorAddr,
+		types.ModuleName,
+		sdk.NewCoins(repayment),
+	); err != nil {
+		return err
+	}
+
+	// Store the remaining borrowed amount in keeper
+	owed := borrowed.AmountOf(repayment.Denom).Sub(repayment.Amount)
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateLoanKey(borrowerAddr, repayment.Denom)
+	if owed.IsZero() {
+		store.Delete(key) // Completely repaid
+	} else {
+		bz, err := owed.Marshal()
+		if err != nil {
+			return err
+		}
+		store.Set(key, bz) // Partially repaid
+	}
+
+	// Transfer uToken collateral reward from borrower to liquidator
+	if err := k.bankKeeper.SendCoins(ctx, borrowerAddr, liquidatorAddr, sdk.NewCoins(reward)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// LiquidationParams computes dynamic liquidation parameters based on collateral denomination,
+// borrowed value, and borrow limit
```

Comment:
```suggestion
// borrowed value, and borrow limit.
```

Reply:
Can we describe what we're actually returning since we're turning multiple values of the same type.

Reply:
Godoc updated for this function